maps/src/MapBinner.cxx

```diff
@@ -0,0 +1,290 @@
+
+	if (T_->GetPolConv() != G3SkyMap::ConvNone) {
+		Q_ = stub_map.Clone(false);
+		Q_->pol_type = G3SkyMap::Q;
+		U_ = stub_map.Clone(false);
+		U_->pol_type = G3SkyMap::U;
+	}
+}
+
+void
+MapBinner::Process(G3FramePtr frame, std::deque<G3FramePtr> &out)
+{
+	if (frame->Has(boloprops_name_))
+		boloprops_ = frame->Get<BolometerPropertiesMap>(
+		    boloprops_name_);
+
+	if (frame->type == G3Frame::EndProcessing) {
+		G3FramePtr out_frame(new G3Frame(G3Frame::Map));
+		out_frame->Put("Id", G3StringPtr(new G3String(output_id_)));
+		out_frame->Put("T",
+		    boost::dynamic_pointer_cast<G3FrameObject>(T_));
+		if (Q_) {
+			out_frame->Put("Q",
+			    boost::dynamic_pointer_cast<G3FrameObject>(Q_));
+			out_frame->Put("U",
+			    boost::dynamic_pointer_cast<G3FrameObject>(U_));
+			out_frame->Put("Wpol", map_weights_);
+		} else {
+			out_frame->Put("Wunpol", map_weights_);
+		}
+		out.push_back(out_frame);
+		out.push_back(frame);
+		return;
+	}
+
+	if (frame->type != G3Frame::Scan) {
+		out.push_back(frame);
+		return;
+	}
+
+	if (!boloprops_)
+		log_fatal("Need bolometer properties before detector data "
+		    "can be processed.");
+
+	G3VectorQuatConstPtr pointing =
+	    frame->Get<G3VectorQuat>(pointing_);
+	if (!pointing) {
+		log_error("Missing pointing %s", pointing_.c_str());
+		out.push_back(frame);
+		return;
+	}
+
+	G3TimestreamMapConstPtr timestreams =
+	    frame->Get<G3TimestreamMap>(timestreams_);
+	G3MapDoubleConstPtr weights = frame->Get<G3MapDouble>(weights_);
+	if (!timestreams) {
+		log_error("Missing timestreams %s", timestreams_.c_str());
+		out.push_back(frame);
+		return;
+	}
+	if (!weights) {
+		log_error("Missing weights %s", weights_.c_str());
+		out.push_back(frame);
+		return;
+	}
+
+	if (!units_set_) {
+		T_->units = timestreams->GetUnits();
+		if (Q_)
+			Q_->units = timestreams->GetUnits();
+		if (U_)
+			U_->units = timestreams->GetUnits();
+		units_set_ = true;
+	} else {
+		if (T_->units != timestreams->GetUnits())
+			log_fatal("Timestreams have units that do not match "
+			    "earlier timestreams in the pipeline.");
+	}
+
+#ifdef OPENMP_FOUND
+	// G3SkyMap is not thread-safe when sparse. Avoid the cost of locks
+	// by making one set of sparse maps per thread per scan (which use
+	// essentially zero memory) and then co-adding them at the end of
+	// the scan.
+	G3SkyMapPtr scanT, scanQ, scanU;
+	G3SkyMapWeightsPtr scanW;
+
+	struct chunk {
+		G3SkyMapPtr scanT, scanQ, scanU;
+		G3SkyMapWeightsPtr scanW;
+	};
+	struct chunk *chunk_array;
+	size_t nchunks;
+
+	// Create a list of detectors to satisfy OpenMP's need for scalar
+	// iteration
+	std::vector<std::string> dets;
+	for (auto i : *timestreams)
+		dets.push_back(i.first);
+
+	// Clamp num_threads to prevent memory balloon?
+	#pragma omp parallel private(scanT, scanQ, scanU, scanW) shared(chunk_array,nchunks)
+	{
+		#pragma omp single
+		{
+			chunk_array = new struct chunk[omp_get_num_threads()];
+			nchunks = omp_get_num_threads();
```

Comment:
I think `omp_get_num_threads()` does what you expect outside of the the omp parallel block.  Why can't `chunk_array` and `nchunks` be initialized at lines 163-164?

Reply:
No, it's OK. It's all part of the same parallel block as below, so omp_get_num_threads() returns (spec'ed and tested) the number of threads participating in the entire parallel block, not just this section.